qcodes/utils/dataset/doNd.py

```diff
@@ -36,34 +37,59 @@ class UnsafeThreadingException(Exception):
     pass
 
 
-def _check_threadsafe(param_meas: Sequence[ParamMeasT]) -> None:
+class _ParamCaller:
 
-    real_parameters = [param for param in param_meas if isinstance(param, _BaseParameter)]
+    def __init__(self, *parameters: _BaseParameter):
 
-    insts = [param.root_instrument for param in real_parameters if param.root_instrument]
-    if len(set(insts)) != len(insts):
-        duplicates = [param for param in real_parameters
-                      if param.root_instrument and insts.count(param.root_instrument) > 1]
-        raise UnsafeThreadingException('Can not use threading to '
-                                       'read '
-                                       'several things from the same '
-                                       'instrument. Specifically, you '
-                                       'asked for'
-                                       ' {}.'.format(duplicates))
+        self._parameters = parameters
+
+    def __call__(self) -> Tuple[ParamDataType, ...]:
+        output = []
+        for param in self._parameters:
+            output.append(param.get())
+            output.append(param.get())
+        return tuple(output)
+
+    def __repr__(self) -> str:
+        names = tuple(param.full_name for param in self._parameters)
+        return f"ParamCaller of {names}"
+
+
+def _instrument_to_param(
+        params: Sequence[ParamMeasT]
+) -> Dict[Optional[str], Tuple[_BaseParameter, ...]]:
+
+    real_parameters = [param for param in params if isinstance(param, _BaseParameter)]
+
+    output: Dict[Optional[str], Tuple[_BaseParameter, ...]] = defaultdict(tuple)
+    for param in real_parameters:
+        if param.root_instrument:
+            output[param.root_instrument.name] += (param,)
+        else:
+            output[None] += (param,)
+
+    return output
 
 
 def _call_params_threaded(param_meas: Sequence[ParamMeasT]) -> OutType:
 
+    inst_param_mapping = _instrument_to_param(param_meas)
+    executors = tuple(_ParamCaller(*param_list)
+                      for param_list in
+                      inst_param_mapping.values())
+
     output: OutType = []
-    threads = [RespondingThread(target=param)
-               for param in param_meas]
+    threads = [RespondingThread(target=executor)
+               for executor in executors]
 
     for t in threads:
         t.start()
 
-    for t, param in zip(threads, param_meas):
-        if isinstance(param, _BaseParameter):
-            output.append((param, t.output()))
+    for t, executor, param_list in zip(threads, executors, inst_param_mapping.values()):
+        thread_output = t.output()
+        assert thread_output is not None
```

Comment:
Is this meant to stay here?

Reply:
If it was None we could not iterate over it (as we do below) and mypy will point that out if you remove the assert. 
Adding an assert is the simplest way to make mypy happy.

Reply:
In practice this cannot be none (at the most it could be a list of None) but it's not simple to make the type system see that without some significant changes to the design of RespondingThread

Reply:
I see. And, in general, instrument drivers do not return None? I'm worried about an overloaded DMM or something causing a problem here.

Reply:
Each parameter could return None just fine. What we are iteration over here is thread.output which is the output of _ParamCaller() which is Tuple[Output_from_param_call, ...] 

The only reason mypy thinks that this variable might be none is that RespondingThread._output is set to None before RespondingThread.run is called